### ISSUE TITLE
How do you get the ansible playbooks onto the boxes in opsworks?

### ISSUE THREAD

**Issue by andyofb** [MyPureCloud/ansible-opsworks#6]
How do you get the ansible playbooks onto the boxes. If you need to run them before deploy I cant deploy them as an app. What is your recommendation?

**Comment by drewdahlke** [comment]
We're getting them on the machine by adding a provisioners section to our packer script. For example

{
  "variables": {
    "role": "",
    "version": "",
    "sourceami": "",
    "playbook_dir": "",
    "chef-dir":"",
    "playbook_file_path": "",
    "inventory_file_path": "",
    "file_provisioner_source": ""
  },
 "provisioners": [
    {
      "type": "file",
      "source": "{{user `file_provisioner_source`}}",
      "destination": "/home/ec2-user"
    },
    {
      "type": "file",
      "source": "{{user `chef-dir`}}",
      "destination": "/home/ec2-user"
    },
    {
      "type": "ansible-local",
      "playbook_file": "{{user `playbook_file_path`}}",
      "extra_arguments": "--verbose --limit localhost -e \"role={{user `role`}} version={{user `version`}} sourceami={{user `sourceami`}}\"",
      "inventory_file": "{{user `inventory_file_path`}}",
      "playbook_dir": "{{user `playbook_dir`}}"
    }
  ]

In our case those variables get populated by a jenkins job which kicks off the packer script to bake a new AMI.

**Comment by andyofb** [comment]
So you build an AMI from a job in jenkins which is then in turn used in opsworks when provisioning a new server?

**Comment by gnethercutt** [comment]
Exactly. We also take the same approach (immutable system images) for the deploys we do with [Asgard](https://github.com/Netflix/asgard).

**Comment by drewdahlke** [comment]
Correct. We've been working on a python script you can use in a jenkins job to take that newly baked AMI and rolling upgrade all the machines in an opsworks stack. That script isn't open source *yet*, but email me and I'm happy to share. 

The script can also be used for bootstrapping the cluster with some instances. When bootstrapping a new cluster with machines for the 1st time you'll want to use a script instead of doing it with cloud formation. The reason being that as soon as you modify a resource that cloud formation created, the CF template can no longer be used to update the stack (eg change security groups). So use CF to create your stack & layers, but manage the instances in the cluster separately. You don't want to be tweaking your CF template to add/remove instances b/c that's not making good use of opsworks.

**Comment by andyofb** [comment]
Sounds like a good solution. I would be interested in seeing it. Thanks for the information.

**Comment by rikh42** [comment]
Could you also add a remote_file resource to the chef recipes that fetch the Ansible scripts from somewhere (github, s3, whatever) just before they are needed...?
